### ISSUE TITLE
minetest.deserialize fails for serialized data containing NaN or infinities

### ISSUE THREAD

**Issue by appgurueu** [minetest/minetest#11899]
##### Minetest version
Any

##### Summary
The way `minetest.serialize` currently formats numbers, NaN will be turned into `nan` or `-nan`, depending on the sign; infinities will be turned into `-inf` or `inf`. The environment `minetest.deserialize` provides however contains neither `nan` nor `inf` variables. Both will therefore evaluate to `nil`. If they are negated, this will lead to a crash (`attempt to negate a nil value`). It will also lead to a crash if they are used as a table index; this isn't an issue for NaN, as it can't occur as a table index anyways, but tables containing positive infinity as a key will lead to an `table index is nil`. The `pcall` will catch this, but the returned object will be `nil` if an error has occurred.

In a nutshell:

* Negative NaN or negative infinity anywhere within the data as well as positive infinity as a table key will cause deserialization to fail;
* Positive infinity or positive NaN elsewhere will be turned into `nil`.

##### Steps to reproduce
The below assertions will all fail:
```lua
assert(minetest.deserialize(minetest.serialize(0/0)))
assert(minetest.deserialize(minetest.serialize(math.huge)))
assert(minetest.deserialize(minetest.serialize({[math.huge] = 0})))
```
